### PR TITLE
Port volume booster to web/vesktop

### PR DESCRIPTION
**Pull request by sadan4**
resolves: #1944 
ports volume booster to web/vesktop

TODO: 

- [x] don't error on screenshare
- [x] adjust volume?
- [x] fix deafening

### CODE REVIEW

**Review by Sqaaakoi — COMMENTED** [review]
works but idk if it really is 200% volume sounds like the "fix" you told me you made where it was 2x the volume it was supposed to be is actually what is intended

tl;dr: i think your fix broke it more

**Comment by Sqaaakoi** [comment]
> It does raise the volume but it doesnt _seem_ to be ex: 500% louder.

yeah this was my experience too

> also it basically functions as a fake deafen at the moment.,...

we cannot ship that, theres a good reason why fake deafen is explicitly not a feature, regardless of that I think I'd still like to have a "I want my friends to be quiet for a bit" button

**Comment by sadan4** [comment]
> also it basically functions as a fake deafen at the moment.,...
somehow I didn't notice this, thanks!

File: src/plugins/volumeBooster/index.ts [hunk 1]
```diff
@@ -28,12 +29,32 @@ const settings = definePluginSettings({
...
         stickToMarkers: true,
-    }
+    },
 });
-
+// for some godforsaken reason, the volume is ran through this formula before its stored. pathcing it out does not work.
+const VolumeEncoder = {
+    decode: findByCodeLazy("6+1:"),
+    encode: findByCodeLazy("50+50"),
+};
```
Comment (Vendicated) [inline src/plugins/volumeBooster/index.ts:38]:
```suggestion
const PerceptualVolume = {
    amplitudeToPerceptual: findByCodeLazy("6+1:"),
    perceptualToAmplitude: findByCodeLazy("50-50"),
};
```

Comment (sadan4) [reply src/plugins/volumeBooster/index.ts]: done

**Comment by Nuckyz** [comment]
You are doing two webpack searches yet only one is used, remove the other one

also the code style doesn't match the rest of project, mainly the spacing in if conditions

File: src/plugins/volumeBooster/index.ts [hunk 2]
```diff
@@ -47,10 +68,33 @@ export default definePlugin({
...
+                    replace: "Math.round(arguments[0])"
+                },
+                // to update the volume on user join
+                {
+                    match: /,this.stream.getTracks\(\).length/,
+                    replace: ",this.updateAudioElement()$&"
+                },
+                // to actually patch the volume
+                {
+                    match: /volume=t.*?;/,
```
[the remark this answers is not in the document]
Comment (sadan4) [reply src/plugins/volumeBooster/index.ts:93]: it is matching the t in `this`

**Comment by sadan4** [comment]
> You are doing two webpack searches yet only one is used, remove the other one
> 
> also the code style doesn't match the rest of project, mainly the spacing in if conditions

I may be a tad stupid, but what are/were you referring to by the spacing in if conditions

File: src/plugins/volumeBooster/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vendicated) [reply src/plugins/volumeBooster/index.ts:93]:
```suggestion
                    match: /\.volume=this._volume/100;/,
```

[the remark this answers is not in the document]
Comment (sadan4) [reply src/plugins/volumeBooster/index.ts]: fixed

File: src/plugins/volumeBooster/index.ts [hunk 3]
```diff
@@ -83,4 +127,17 @@ export default definePlugin({
             ]
         }
     ],
+    patchVolume(data: StreamData){
+        if(data.stream.getAudioTracks().length === 0) return;
```
Comment (Vendicated) [inline src/plugins/volumeBooster/index.ts:131]:
```suggestion
        if (data.stream.getAudioTracks().length === 0) return;
```

File: src/plugins/volumeBooster/index.ts [hunk 4]
```diff
@@ -83,4 +127,17 @@ export default definePlugin({
...
+
+        data.streamSourceNode ??= data.audioContext.createMediaStreamSource(data.stream);
+
+        if(!data.gainNode){
```
Comment (Vendicated) [inline src/plugins/volumeBooster/index.ts:135]:
```suggestion
        if (!data.gainNode){
```

File: src/plugins/volumeBooster/index.ts [hunk 5]
```diff
@@ -83,4 +127,17 @@ export default definePlugin({
             ]
         }
     ],
+    patchVolume(data: StreamData){
```
Comment (Vendicated) [inline src/plugins/volumeBooster/index.ts:130]:
```suggestion
    patchVolume(data: StreamData) {
```

File: src/plugins/volumeBooster/index.ts [hunk 6]
```diff
@@ -47,10 +68,33 @@ export default definePlugin({
...
+                      + `:${minorMaxVolume}*$self.settings.store.multiplier`
             }
         })),
+        // PATCHES NEEDED FOR WEB/VESKTOP
+        {
+            find: "streamSourceNode",
+            predicate: () => !IS_DISCORD_DESKTOP,
+            group: true,
+            replacement: [
+                // to pervent the cap of 100
```
Comment (Vendicated) [inline src/plugins/volumeBooster/index.ts:81]:
```suggestion
                // remove the cap of 100%
```

File: src/plugins/volumeBooster/index.ts [hunk 7]
```diff
@@ -28,12 +29,32 @@ const settings = definePluginSettings({
...
-    }
+    },
 });
-
+// for some godforsaken reason, the volume is ran through this formula before its stored. pathcing it out does not work.
+const PerceptualVolume = {
+    amplitudeToPerceptual: findByCodeLazy("6+1:"),
+    // perceptualToAmplitude: findByCodeLazy("50-50"),
+};
+interface StreamData{
```
Comment (Vendicated) [inline src/plugins/volumeBooster/index.ts:39]:
```suggestion
interface StreamData {
```

**Review by Vendicated — COMMENTED** [review]
like nuckyz said, follow the code style

File: src/plugins/volumeBooster/index.ts [hunk 8]
```diff
@@ -28,12 +29,32 @@ const settings = definePluginSettings({
         markers: makeRange(1, 5, 1),
         default: 2,
         stickToMarkers: true,
-    }
+    },
```
Comment (Nuckyz) [inline src/plugins/volumeBooster/index.ts:32]:
```suggestion
    }
```

File: src/plugins/volumeBooster/index.ts [hunk 9]
```diff
@@ -28,12 +29,32 @@ const settings = definePluginSettings({
...
 });
-
+// for some godforsaken reason, the volume is ran through this formula before its stored. pathcing it out does not work.
+const PerceptualVolume = {
+    amplitudeToPerceptual: findByCodeLazy("6+1:"),
+    // perceptualToAmplitude: findByCodeLazy("50-50"),
+};
```
Comment (Nuckyz) [inline src/plugins/volumeBooster/index.ts:38]:
```suggestion
const perceptualToAmplitude = findByCodeLazy("50-50");

```

File: src/plugins/volumeBooster/index.ts [hunk 10]
```diff
@@ -83,4 +127,17 @@ export default definePlugin({
             ]
         }
     ],
+    patchVolume(data: StreamData) {
```
Comment (Nuckyz) [inline src/plugins/volumeBooster/index.ts:130]:
```suggestion

    patchVolume(data: StreamData) {
```

**Review by Nuckyz — CHANGES_REQUESTED** [review]

File: src/plugins/volumeBooster/index.ts [hunk 11]
```diff
@@ -83,4 +127,17 @@ export default definePlugin({
...
+
+        data.streamSourceNode ??= data.audioContext.createMediaStreamSource(data.stream);
+
+        if (!data.gainNode){
+            const gain = data.gainNode = data.audioContext.createGain();
+            data.streamSourceNode.connect(gain);
+            gain.connect(data.audioContext.destination);
+        }
+
+        data.gainNode.gain.value = data._mute ? 0 : PerceptualVolume.amplitudeToPerceptual(data._volume)/100;
```
Comment (Nuckyz) [inline src/plugins/volumeBooster/index.ts:141]:
```suggestion
        data.gainNode.gain.value = data._mute ? 0 : amplitudeToPerceptual(data._volume)/100;
```

**Review by Nuckyz — CHANGES_REQUESTED** [review]

File: src/plugins/volumeBooster/index.ts (hunk 9 above)
Comment (sadan4) [reply src/plugins/volumeBooster/index.ts:38]: I'll push the fixes to your changes when I push my fix for low volumes

**Comment by Vendicated** [comment]
i think it would be a good idea to extract this into a separate plugin like WebExtraVolume and make it enabled by default. That way more users can profit from it and Vesktop users will get the expected stock equivalent behaviour

**Comment by sadan4** [comment]
> i think it would be a good idea to extract this into a separate plugin like WebExtraVolume and make it enabled by default. That way more users can profit from it and Vesktop users will get the expected stock equivalent behaviour

It depends on the patches already introduced by Volume Booster.

**Comment by MaxBosse** [comment]
Is there anything I can help with to get this merged? :) Would love to have this feature on web/vesktop

**Comment by sadan4** [comment]
This should be ready as i fixed the stock discord issue that made it glitch out at low volumes

**Comment by Nuckyz** [comment]
Thank you!

**Comment by sadan4** [comment]
Apparently this isn't working on the vesktop, which is weird because that's where I tested it. I have family over today, but I will work on it tomorrow if nobody else fixes it

**Pull request merged**